hydra-node/src/Hydra/Chain/Direct/Handlers.hs

```diff
@@ -0,0 +1,418 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Provide infrastructure-independent "handlers" for posting transactions and following the chain.
+--
+-- This module encapsulates the transformation logic between cardano transactions and `HydraNode` abstractions
+-- `PostChainTx` and `OnChainTx`, and maintainance of on-chain relevant state.
+module Hydra.Chain.Direct.Handlers where
+
+import Hydra.Prelude
+
+import Cardano.Ledger.Alonzo.Tx (ValidatedTx)
+import Cardano.Ledger.Alonzo.TxSeq (txSeqTxns)
+import Cardano.Ledger.Crypto (StandardCrypto)
+import Cardano.Ledger.Shelley.API (TxId)
+import qualified Cardano.Ledger.Shelley.API as Ledger
+import Control.Monad (foldM)
+import Control.Monad.Class.MonadSTM (
+  readTVarIO,
+  retry,
+  writeTVar,
+ )
+import Control.Monad.Class.MonadTimer (timeout)
+import Data.Aeson (Value (String), object, (.=))
+import Data.Sequence.Strict (StrictSeq)
+import Hydra.Cardano.Api (
+  ChainPoint (..),
+  PaymentKey,
+  SlotNo,
+  Tx,
+  VerificationKey,
+  fromConsensusPointHF,
+  fromLedgerTx,
+  fromLedgerTxIn,
+  fromLedgerUTxO,
+  toLedgerTx,
+  toLedgerUTxO,
+ )
+import Hydra.Chain (
+  Chain (..),
+  ChainEvent (..),
+  OnChainTx (..),
+  PostChainTx (..),
+  PostTxError (..),
+ )
+import Hydra.Chain.Direct.State (
+  SomeOnChainHeadState (..),
+  TokHeadState (..),
+  abort,
+  close,
+  collect,
+  commit,
+  contest,
+  fanout,
+  getContestationDeadline,
+  getKnownUTxO,
+  initialize,
+  observeSomeTx,
+  reifyState,
+ )
+import Hydra.Chain.Direct.Tx (PointInTime)
+import Hydra.Chain.Direct.Util (
+  Block,
+  Era,
+  SomePoint (..),
+ )
+import Hydra.Chain.Direct.Wallet (
+  ErrCoverFee (..),
+  TinyWallet (..),
+  TinyWalletLog,
+  getFuelUTxO,
+  getTxId,
+ )
+import Hydra.Data.ContestationPeriod (posixToUTCTime)
+import Hydra.Logging (Tracer, traceWith)
+import Ouroboros.Consensus.Cardano.Block (HardForkBlock (BlockAlonzo))
+import Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock (..))
+import Ouroboros.Network.Block (Point (..), blockPoint)
+import Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+
+--
+
+-- * Posting Transactions
+
+--
+
+-- | A simple interface used by `Chain` to actually post the transaction to the chain.
+-- This function can be synchronous or asynchronous. It `atomically` executes
+-- the given `STM` action and takes care of posting the resulting transaction to
+-- the chain.
+type TxPoster m = STM m (ValidatedTx Era) -> m ()
```

Comment:
I tried but this would be a larger change: The STM action is wrapped inside another STM action in the called in order to provide a synchronous interface, by creating and passing the `TMVar` to hold the result. I tried to change this logic to something simpler but quite a few tests fail and did not want to spend the time to investigate.